core/controller/src/main/scala/whisk/core/controller/Meta.scala

```diff
@@ -16,23 +16,135 @@
+        "text" -> resultAsText _,
+        "js" -> resultAsJson _,
+        "http" -> resultAsHttp _)
+
+    val supportedMediaTypes = mediaTranscoders.keySet
+
+    def metadata(method: String, headers: String, path: String, namespace: Option[String]) = {
+        Map("__ow_meta_verb" -> method.toLowerCase.toJson,
+            "__ow_meta_headers" -> headers.toJson,
+            "__ow_meta_path" -> path.toJson) ++
+            namespace.map(ns => "__ow_meta_namespace" -> ns.toJson)
+    }
+
+    protected def resultAsHtml(result: JsValue): RequestContext => Unit = {
+        Try {
+            val JsString(str) = result
+            val input = new ByteArrayInputStream(str.getBytes)
+            scala.xml.XML.load(input)
+        } match {
+            case Success(xml) => respondWithMediaType(`text/html`) { complete(OK, xml) }
+            case Failure(t)   => complete(BadRequest, Messages.invalidMedia(`text/html`))
+        }
+    }
+
+    protected def resultAsText(result: JsValue): RequestContext => Unit = {
+        result match {
+            case r: JsObject  => complete(OK, r.compactPrint)
+            case r: JsArray   => complete(OK, r.compactPrint)
+            case JsString(s)  => complete(OK, s)
+            case JsBoolean(b) => complete(OK, b.toString)
+            case JsNumber(n)  => complete(OK, n.toString)
+            case JsNull       => complete(OK)
+        }
+    }
+
+    protected def resultAsJson(result: JsValue): RequestContext => Unit = {
+        result match {
+            case r: JsObject  => complete(OK, r)
+            case r: JsArray   => complete(OK, r)
+            case JsString(s)  => complete(OK, s)
+            case JsBoolean(b) => complete(OK, b.toString)
+            case JsNumber(n)  => complete(OK, n.toString)
+            case JsNull       => complete(OK)
+        }
+    }
+
+    protected def resultAsHttp(result: JsValue): RequestContext => Unit = {
+        result match {
+            case JsObject(fields) => Try {
+                val headers = fields.get("headers").map {
+                    case JsObject(hs) => hs.map {
+                        case (k, JsString(v))  => RawHeader(k, v)
+                        case (k, JsBoolean(v)) => RawHeader(k, v.toString)
+                        case (k, JsNumber(v))  => RawHeader(k, v.toString)
+                        case _                 => throw new Throwable("Invalid header")
+                    }.toList
+
+                    case _ => throw new Throwable("Invalid header")
+                } getOrElse List()
+
+                val code = fields.get("code").map {
+                    case JsNumber(c) =>
+                        val status = c.toIntExact // throws exception if not whole
+                        if (status >= 100 && status < 600) {
+                            status
+                        } else throw new Throwable("Illegal code")
+
+                    case _ => throw new Throwable("Illegal code")
+                } getOrElse 200
```

Comment:
You might want to check out https://github.com/spray/spray/blob/master/spray-http/src/main/scala/spray/http/StatusCode.scala#L29-L35 which also throws an `Exception` on an illegal code and does all the standard handling for you.